File: backend/backstop/core/attribution.py
"""Attribution: can a difference between two runs be pinned on one change?

A comparison is only as good as its isolation. Two runs may differ in the prompt,
the model (or how it was executed), the rule date, the calls scored, the contract
set, the judge policy and the approved overrides in force. When exactly one of
those differs, every difference in the results is attributable to it (subject to
the significance test). When two or more differ, the comparison is confounded:
the numbers are still true, but no single change can be credited or blamed.

For a confounded pair this module also searches the runs that already exist for
pairs that change one factor at a time, so a reviewer is pointed at a clean
comparison instead of being left with a caveat.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from backstop.models import Run

# Not fingerprinted yet (documented in docs/honesty.md): generation settings (temperature,
# max tokens, tool vs JSON mode), the simulated defect profile, the Ollama model digest,
# and which approved overrides are in scope (only their number is compared).

# Order is the order a reader scans them in: what was asked, of whom, under which rules, on what.
FACTORS: tuple[str, ...] = ("prompt", "model", "rule_date", "corpus", "contract_set", "judge", "overrides")

LABELS: dict[str, str] = {
    "prompt": "prompt",
    "model": "model",
    "rule_date": "rule date",
    "corpus": "calls scored",
    "contract_set": "contract set",
    "judge": "judge",
    "overrides": "approved overrides",
}

# Factors that reach only the judged (advisory) contracts; a change confined to them
# leaves the deterministic, release-blocking verdicts attributable to the other change.
JUDGED_ONLY: frozenset[str] = frozenset({"judge"})


@dataclass(frozen=True)
class RunFactors:
    """The value of every factor for one run: `key` is compared, `display` is shown."""

    run_id: str
    key: dict[str, str]
    display: dict[str, str]


def factors_of(run: Run) -> RunFactors:
    stats: dict[str, Any] = run.stats or {}
    model_id = run.model.model_id
    judge = stats.get("judge_model_id") or model_id
    judge_policy = "same model as the run" if judge == model_id else judge
    judge_n = stats.get("judge_n")
    overrides = int((stats.get("test_cases") or {}).get("in_scope", 0))
    corpus_name = stats.get("corpus", "synthetic")
    calls = stats.get("transcripts")
    key = {
        "prompt": run.prompt_version_id,
        "model": f"{model_id}|{run.adapter}",
        "rule_date": run.rule_date.isoformat(),
        "corpus": run.corpus_hash,
        "contract_set": run.contract_set_hash,
        "judge": f"{judge_policy}|{judge_n}" if judge_n else judge_policy,
        "overrides": str(overrides),
    }
    display = {
        "prompt": f"v{run.prompt_version.version}",
        "model": f"{model_id} ({run.adapter})",
        "rule_date": run.rule_date.isoformat(),
        "corpus": f"{corpus_name}, {calls} calls" if calls else corpus_name,
        "contract_set": run.contract_set_hash[:8],
        "judge": judge_policy,
        "overrides": f"{overrides} in scope",
    }
    return RunFactors(run_id=run.id, key=key, display=display)


# Factors that follow from another: the approved overrides in scope are the ones that
# touch the calls scored, so a different set of calls brings different overrides with it.
DERIVED_FROM: dict[str, str] = {"overrides": "corpus"}


def changed_factors(a: RunFactors, b: RunFactors) -> list[str]:
    changed = [f for f in FACTORS if a.key[f] != b.key[f]]
    return [f for f in changed if DERIVED_FROM.get(f) not in changed]


def _join(labels: list[str]) -> str:
    if len(labels) <= 2:
        return " and ".join(labels)
    return ", ".join(labels[:-1]) + " and " + labels[-1]


def isolating_pairs(a: RunFactors, b: RunFactors, candidates: Iterable[RunFactors]) -> list[dict[str, str]]:
    """For each factor that differs between A and B, an existing pair of runs (C, D) that
    differs in that factor alone and moves it the same way: C has A's value, D has B's.

    The other factors may sit at any shared value, so a confounded jump from A to B can
    decompose into a chain of clean steps (rule date, then prompt, then model). Pairs
    that include A or B are preferred; otherwise the newest pair wins (candidates are
    expected newest first).
    """
    pool: list[RunFactors] = [a, b] + [c for c in candidates if c.run_id not in (a.run_id, b.run_id)]
    out: list[dict[str, str]] = []
    for factor in changed_factors(a, b):
        best: tuple[int, int, str, str] | None = None
        for i, c in enumerate(pool):
            if c.key[factor] != a.key[factor]:
                continue
            for j, d in enumerate(pool):
                if d.key[factor] != b.key[factor] or changed_factors(c, d) != [factor]:
                    continue
                anchored = 0 if {c.run_id, d.run_id} & {a.run_id, b.run_id} else 1
                rank = (anchored, i + j, c.run_id, d.run_id)
                if best is None or rank[:2] < best[:2]:
                    best = rank
        if best is not None:
            out.append({"factor": factor, "label": LABELS[factor], "a_run_id": best[2], "b_run_id": best[3]})
    return out


def attribute(a: RunFactors, b: RunFactors, candidates: Iterable[RunFactors] = ()) -> dict[str, Any]:
    """Verdict, the factors that changed (with both values), what was held constant, and a
    plain-English summary a reviewer can act on."""
    changed = changed_factors(a, b)
    rows = [{"factor": f, "label": LABELS[f], "a": a.display[f], "b": b.display[f]} for f in changed]
    held = [LABELS[f] for f in FACTORS if f not in changed]
    labels = [LABELS[f] for f in changed]
    scope_note: str | None = None
    suggestions: list[dict[str, str]] = []

    if not changed:
        verdict = "repeat"
        summary = ("No tracked input differs between A and B. A difference comes from run-to-run variation "
                   "(which the significance test below weighs) or from something Backstop does not yet "
                   "fingerprint: generation settings, a simulated defect profile, or model weights re-pulled "
                   "under the same tag.")
    elif len(changed) == 1:
        verdict = "isolated"
        f = changed[0]
        summary = (f"Only the {LABELS[f]} changed ({a.display[f]} → {b.display[f]}). Every other tracked input "
                   f"was held constant, so the differences below are attributable to that change, subject to "
                   f"the significance test.")
    else:
        verdict = "confounded"
        suggestions = isolating_pairs(a, b, candidates)
        summary = (f"The {_join(labels)} changed together. The differences below are real, but they cannot "
                   f"be attributed to any one of these changes.")
        found = [s["label"] for s in suggestions]
        missing = [LABELS[f] for f in changed if LABELS[f] not in found]
        if found and not missing:
            summary += " Existing runs isolate each change on its own; compare those instead."
        elif found:
            summary += (f" Existing runs isolate the {_join(found)} change{'s' if len(found) > 1 else ''}; no pair "
                        f"isolates the {_join(missing)} change{'s' if len(missing) > 1 else ''} yet.")
        else:
            summary += " No existing pair of runs isolates them; start runs that change one thing at a time."
        primary = [f for f in changed if f not in JUDGED_ONLY]
        if len(primary) == 1:
            scope_note = (f"Deterministic, release-blocking contracts are still attributable to the "
                          f"{LABELS[primary[0]]} change; only the judged (advisory) contracts mix in the "
                          f"{_join([LABELS[f] for f in changed if f in JUDGED_ONLY])} change.")

    return {"verdict": verdict, "changed": rows, "held_constant": held, "summary": summary,
            "scope_note": scope_note, "isolating_pairs": suggestions}

File: frontend/src/lib/attribution.ts
/**
 * Attribution, client side: the same rules as backend/backstop/core/attribution.py,
 * used by mock mode so the backend-free console answers "can this difference be
 * pinned on one change?" exactly as the server does.
 */
import type { AttributionOut, FactorChangeOut, IsolatingPairOut, RunOut } from '../api/types';

export const FACTORS = ['prompt', 'model', 'rule_date', 'corpus', 'contract_set', 'judge', 'overrides'] as const;
export type Factor = (typeof FACTORS)[number];

export const FACTOR_LABELS: Record<Factor, string> = {
  prompt: 'prompt',
  model: 'model',
  rule_date: 'rule date',
  corpus: 'calls scored',
  contract_set: 'contract set',
  judge: 'judge',
  overrides: 'approved overrides',
};

/** Factors that reach only the judged (advisory) contracts. */
const JUDGED_ONLY: ReadonlySet<Factor> = new Set(['judge']);
/** Overrides in scope follow from the calls scored. */
const DERIVED_FROM: Partial<Record<Factor, Factor>> = { overrides: 'corpus' };

interface RunFactors {
  id: string;
  key: Record<Factor, string>;
  display: Record<Factor, string>;
}

function num(v: unknown): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : 0;
}

export function factorsOf(run: RunOut): RunFactors {
  const stats = (run.stats ?? {}) as Record<string, unknown>;
  const judgeId = typeof stats.judge_model_id === 'string' && stats.judge_model_id ? stats.judge_model_id : run.model_id;
  const judge = judgeId === run.model_id ? 'same model as the run' : judgeId;
  const testCases = (stats.test_cases ?? {}) as Record<string, unknown>;
  const overrides = String(num(testCases.in_scope));
  const corpusName = String(run.corpus ?? stats.corpus ?? 'synthetic');
  const calls = num(stats.transcripts);
  return {
    id: run.id,
    key: {
      prompt: run.prompt_hash,
      model: `${run.model_id}|${run.adapter}`,
      rule_date: run.rule_date,
      corpus: run.corpus_hash,
      contract_set: run.contract_set_hash,
      judge,
      overrides,
    },
    display: {
      prompt: `v${run.prompt_version}`,
      model: `${run.model_id} (${run.adapter})`,
      rule_date: run.rule_date,
      corpus: calls ? `${corpusName}, ${calls} calls` : corpusName,
      contract_set: run.contract_set_hash.slice(0, 8),
      judge,
      overrides: `${overrides} in scope`,
    },
  };
}

export function changedFactors(a: RunFactors, b: RunFactors): Factor[] {
  const changed = FACTORS.filter((f) => a.key[f] !== b.key[f]);
  return changed.filter((f) => {
    const parent = DERIVED_FROM[f];
    return !(parent && changed.includes(parent));
  });
}

function join(labels: string[]): string {
  if (labels.length <= 2) return labels.join(' and ');
  return `${labels.slice(0, -1).join(', ')} and ${labels[labels.length - 1]}`;
}

/** Existing pairs that move one factor alone, from A's value to B's; pairs touching A or B first. */
export function isolatingPairs(a: RunFactors, b: RunFactors, candidates: RunFactors[]): IsolatingPairOut[] {
  const pool = [a, b, ...candidates.filter((c) => c.id !== a.id && c.id !== b.id)];
  const out: IsolatingPairOut[] = [];
  for (const factor of changedFactors(a, b)) {
    let best: { anchored: number; order: number; c: string; d: string } | null = null;
    pool.forEach((c, i) => {
      if (c.key[factor] !== a.key[factor]) return;
      pool.forEach((d, j) => {
        if (d.key[factor] !== b.key[factor]) return;
        const moved = changedFactors(c, d);
        if (moved.length !== 1 || moved[0] !== factor) return;
        const anchored = [c.id, d.id].some((id) => id === a.id || id === b.id) ? 0 : 1;
        if (!best || anchored < best.anchored || (anchored === best.anchored && i + j < best.order)) {
          best = { anchored, order: i + j, c: c.id, d: d.id };
        }
      });
    });
    const found = best as { c: string; d: string } | null;
    if (found) out.push({ factor, label: FACTOR_LABELS[factor], a_run_id: found.c, b_run_id: found.d });
  }
  return out;
}

export function attributeRuns(runA: RunOut, runB: RunOut, others: RunOut[] = []): AttributionOut {
  const a = factorsOf(runA);
  const b = factorsOf(runB);
  const changed = changedFactors(a, b);
  const rows: FactorChangeOut[] = changed.map((f) => ({ factor: f, label: FACTOR_LABELS[f], a: a.display[f], b: b.display[f] }));
  const held = FACTORS.filter((f) => !changed.includes(f)).map((f) => FACTOR_LABELS[f]);
  const labels = changed.map((f) => FACTOR_LABELS[f]);

  if (changed.length === 0) {
    return {
      verdict: 'repeat', changed: rows, held_constant: held, scope_note: null, isolating_pairs: [],
      summary: 'No tracked input differs between A and B. A difference comes from run-to-run variation (which the significance test below weighs) or from something Backstop does not yet fingerprint: generation settings, a simulated defect profile, or model weights re-pulled under the same tag.',
    };
  }
  if (changed.length === 1) {
    const f = changed[0];
    return {
      verdict: 'isolated', changed: rows, held_constant: held, scope_note: null, isolating_pairs: [],
      summary: `Only the ${FACTOR_LABELS[f]} changed (${a.display[f]} → ${b.display[f]}). Every other tracked input was held constant, so the differences below are attributable to that change, subject to the significance test.`,
    };
  }
  const pairs = isolatingPairs(a, b, others.map(factorsOf));
  const found = pairs.map((p) => p.label);
  const missing = labels.filter((l) => !found.includes(l));
  let summary = `The ${join(labels)} changed together. The differences below are real, but they cannot be attributed to any one of these changes.`;
  if (found.length && !missing.length) summary += ' Existing runs isolate each change on its own; compare those instead.';
  else if (found.length) summary += ` Existing runs isolate the ${join(found)} change${found.length > 1 ? 's' : ''}; no pair isolates the ${join(missing)} change${missing.length > 1 ? 's' : ''} yet.`;
  else summary += ' No existing pair of runs isolates them; start runs that change one thing at a time.';
  const primary = changed.filter((f) => !JUDGED_ONLY.has(f));
  const scope_note =
    primary.length === 1
      ? `Deterministic, release-blocking contracts are still attributable to the ${FACTOR_LABELS[primary[0]]} change; only the judged (advisory) contracts mix in the ${join(changed.filter((f) => JUDGED_ONLY.has(f)).map((f) => FACTOR_LABELS[f]))} change.`
      : null;
  return { verdict: 'confounded', changed: rows, held_constant: held, summary, scope_note, isolating_pairs: pairs };
}
